File: LeetCode75_LevelTwo/Most_Stones_Removed_with_Same_Row_or_Column/Most_Stones_Removed_with_Same_Row_or_Column.cc
// 947. Most Stones Removed with Same Row or Column C++
#include <vector>
using namespace std;

class Solution {
public:
    int shareSameRowOrColumn(vector<int>& a, vector<int>& b) {
        return a[0] == b[0] || a[1] == b[1];
    }
    
    // Returns the representative of vertex x
    int find(vector<int>& rep, int x) {
        if (x == rep[x]) {
            return x;
        }
        // Uses Path compression
        return rep[x] = find(rep, rep[x]);
    }
    
    // Combine the stone x and y, and returns 1 if they were not connected
    int performUnion(vector<int>& rep, vector<int>& size, int x, int y) {
        x = find(rep, x);
        y = find(rep, y);
        
        if (x == y) {
            return 0;
        }
        
        if (size[x] > size[y]) {
            rep[y] = x;
            size[x] += size[y];
        } else {
            rep[x] = y;
            size[y] += size[x];
        }
        
        return 1;
    }
    
    int removeStones(vector<vector<int>>& stones) {
        vector<int> rep(stones.size());
        vector<int> size(stones.size());
        // Initialize rep to itself and size as 1
        for (int i = 0; i < stones.size(); i++) {
            rep[i] = i;
            size[i] = 1;
        }
        
        int componentCount = stones.size();
        for (int i = 0; i < stones.size(); i++) {
            for (int j = i + 1; j < stones.size(); j++) {
                if (shareSameRowOrColumn(stones[i], stones[j])) {
                    // Decrement the componenets if union invloved merging
                    componentCount -= performUnion(rep, size, i, j);
                }
            }
        }
        
        // Return the maximum stone that can be removed
        return stones.size() - componentCount;
    }
};

File: LeetCode75_LevelTwo/Most_Stones_Removed_with_Same_Row_or_Column/Most_Stones_Removed_with_Same_Row_or_Column.py
# 947. Most Stones Removed with Same Row or Column Python
from collections import defaultdict

class Solution(object):
    def removeStones(self, stones):
        """
        :type stones: List[List[int]]
        :rtype: int
        """
        def remove_point(a,b):
            points.discard((a,b))
            for y in x_dic[a]:
                if (a,y) in points:
                    remove_point(a,y)

            for x in y_dic[b]:
                if (x,b) in points:
                    remove_point(x,b)
        
        x_dic = defaultdict(list)
        y_dic = defaultdict(list)
        points= {(i,j) for i,j in stones}
        
        for i,j in stones:
            x_dic[i].append(j)
            y_dic[j].append(i)

        cnt = 0
        for a,b in stones:
            if (a,b) in points:
                remove_point(a,b)
                cnt+=1

        return len(stones)-cnt
